File: frontend/src/lib/api/rename.ts
import { apiFetch } from './client';

interface RenameFileEntry {
	id: string;
	path: string;
}

export interface RenamePreview {
	id: string;
	old_name: string;
	new_name: string;
	conflict: boolean;
}

export interface RenameResult {
	id: string;
	status: string;
	old_name: string;
	new_name: string;
	/** Relative path of the file after the rename (new location on success). */
	new_relative_path: string;
	error?: string;
}

export async function previewRenames(
	files: RenameFileEntry[],
	format: string
): Promise<RenamePreview[]> {
	const res = await apiFetch<{ previews: RenamePreview[] }>('/rename/preview', {
		method: 'POST',
		body: JSON.stringify({ files, format })
	});
	return res.previews;
}

export async function executeRenames(
	files: RenameFileEntry[],
	format: string
): Promise<RenameResult[]> {
	const res = await apiFetch<{ results: RenameResult[] }>('/rename/execute', {
		method: 'POST',
		body: JSON.stringify({ files, format })
	});
	return res.results;
}

File: crates/tunewright-server/src/routes/rename.rs
use axum::extract::State;
use axum::Json;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tunewright_core::rename::{self, RenamePreview, RenameResult};
use tunewright_core::scanner;

use crate::error::{join_error, AppError};
use crate::state::AppState;

#[derive(Deserialize)]
pub struct RenameRequest {
    pub files: Vec<RenameFileEntry>,
    pub format: String,
}

#[derive(Deserialize)]
pub struct RenameFileEntry {
    pub id: String,
    pub path: String,
}

#[derive(Serialize)]
pub struct PreviewResponse {
    pub previews: Vec<RenamePreview>,
}

#[derive(Serialize)]
pub struct ExecuteResponse {
    pub results: Vec<RenameResult>,
}

pub async fn preview(
    State(state): State<AppState>,
    Json(body): Json<RenameRequest>,
) -> Result<Json<PreviewResponse>, AppError> {
    let data_root = state.data_root.clone();
    let format = body.format.clone();

    let previews = tokio::task::spawn_blocking(move || {
        let files: Vec<(String, String, PathBuf)> = body
            .files
            .into_iter()
            .filter_map(|f| {
                scanner::resolve_safe_path(&data_root, &f.path)
                    .ok()
                    .map(|safe_path| (f.id, f.path, safe_path))
            })
            .collect();

        rename::preview_renames(&data_root, &files, &format)
    })
    .await
    .map_err(join_error)??;

    Ok(Json(PreviewResponse { previews }))
}

pub async fn execute(
    State(state): State<AppState>,
    Json(body): Json<RenameRequest>,
) -> Result<Json<ExecuteResponse>, AppError> {
    let data_root = state.data_root.clone();
    let format = body.format.clone();

    let results = tokio::task::spawn_blocking(move || {
        let files: Vec<(String, String, PathBuf)> = body
            .files
            .into_iter()
            .filter_map(|f| {
                scanner::resolve_safe_path(&data_root, &f.path)
                    .ok()
                    .map(|safe_path| (f.id, f.path, safe_path))
            })
            .collect();

        rename::execute_renames(&data_root, &files, &format)
    })
    .await
    .map_err(join_error)?;

    Ok(Json(ExecuteResponse { results }))
}
